### PR TITLE
Lvol snapshot management of User Created flag

### PR DESCRIPTION
**Pull request by DamiaSan**
#### Which issue(s) this PR fixes:
Issue #7578

### CODE REVIEW

**Review by shuo-wu — DISMISSED** [review]
The `UserCreated` flag modification LGTM.
But the vendor update introduced xattrs `SnapshotTime` as well. We need to add it for `ReplicaSnapshotCreate` so that one snapshot has the same timestamp among all replicas. Besides, for snapshots, the creation time should be loaded from the xattrs instead.

**Comment by DamiaSan** [comment]
@innobead @derekbit @shuo-wu  I don't have the permission to merge this PR. Who can make it?

**Comment by innobead** [comment]
> @innobead @derekbit @shuo-wu I don't have the permission to merge this PR. Who can make it?

You have `write` permission, so should be able to merge it :thinking: Wait, I will review this later first before merging. Thanks.

File: pkg/spdk/server.go [hunk 1]
```diff
@@ -893,7 +893,11 @@ func (s *Server) EngineSnapshotCreate(ctx context.Context, req *spdkrpc.Snapshot
 		return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find engine %v for snapshot creation", req.Name)
 	}
 
-	snapshotName, err := e.SnapshotCreate(spdkClient, req.SnapshotName)
+	opts := util.SnapshotOptions{
+		UserCreated: req.UserCreated,
+	}
+
+	snapshotName, err := e.SnapshotCreate(spdkClient, req.SnapshotName, opts)
```
Comment (innobead) [inline pkg/spdk/server.go:900]: Why don't we use the same way for both replica and engine snapshot call? The engine call is using an option struct, but the replica call is just using a bool parameter.
> 	return r.SnapshotCreate(spdkClient, req.SnapshotName, req.UserCreated)

Can we just make them consistent?

**Review by innobead — CHANGES_REQUESTED** [review]

File: pkg/spdk/engine.go [hunk 2]
```diff
@@ -1067,21 +1067,21 @@ const (
...
 }
 
 func (e *Engine) SnapshotRevert(spdkClient *spdkclient.Client, snapshotName string) (err error) {
-	_, err = e.snapshotOperation(spdkClient, snapshotName, SnapshotOperationRevert)
+	_, err = e.snapshotOperation(spdkClient, snapshotName, SnapshotOperationRevert, util.SnapshotOptions{})
 	return err
 }
 
-func (e *Engine) snapshotOperation(spdkClient *spdkclient.Client, inputSnapshotName string, snapshotOp SnapshotOperationType) (snapshotName string, err error) {
+func (e *Engine) snapshotOperation(spdkClient *spdkclient.Client, inputSnapshotName string, snapshotOp SnapshotOperationType, opts util.SnapshotOptions) (snapshotName string, err error) {
```
[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/spdk/engine.go]: Done

File: pkg/spdk/replica.go [hunk 3]
```diff
@@ -835,6 +847,9 @@ func (r *Replica) SnapshotCreate(spdkClient *spdkclient.Client, snapshotName str
 		return nil, fmt.Errorf("zero or multiple snap lvols with UUID %s found after lvol snapshot", snapUUID)
 	}
 
+	if userCreatedFlag {
+		bdevLvolList[0].DriverSpecific.Lvol.Xattrs[spdkclient.UserCreated] = "true"
+	}
```
[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/spdk/replica.go]: Done

File: pkg/util/util.go [hunk 4]
```diff
@@ -22,6 +22,10 @@ import (
 	"github.com/longhorn/go-common-libs/proc"
 )
 
+type SnapshotOptions struct {
+	UserCreated bool
+}
```
[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/util/util.go]: Done

File: pkg/spdk/server.go (hunk 1 above)
Comment (DamiaSan) [reply pkg/spdk/server.go]: Done

File: pkg/spdk/engine.go [hunk 5]
```diff
@@ -788,7 +788,7 @@ func (e *Engine) ReplicaAddStart(spdkClient *spdkclient.Client, replicaName, rep
 
 	// TODO: For online rebuilding, the IO should be paused first
 	snapshotName := GenerateRebuildingSnapshotName()
-	updateRequired, err = e.snapshotOperationWithoutLock(spdkClient, replicaClients, snapshotName, SnapshotOperationCreate)
+	updateRequired, err = e.snapshotOperationWithoutLock(spdkClient, replicaClients, snapshotName, SnapshotOperationCreate, nil)
```
[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/spdk/engine.go:791]: If a snapshot doesn't have the user_created xattr, the default is false:
https://github.com/longhorn/longhorn-spdk-engine/blob/d67b39ea8de3a3c24323e261b8b1ee87e293e777/pkg/spdk/replica.go#L162
So it is not needed to explicitly add the flag.

File: pkg/spdk/replica.go [hunk 6]
```diff
@@ -1709,7 +1722,7 @@ func (r *Replica) postFullRestoreOperations(spdkClient *spdkclient.Client, resto
 
 	r.log.Infof("Taking snapshot %v of the restored volume", restore.SnapshotName)
 
-	_, err := r.SnapshotCreate(spdkClient, restore.SnapshotName)
+	_, err := r.SnapshotCreate(spdkClient, restore.SnapshotName, nil)
```
[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/spdk/replica.go:1725]: same as above

File: pkg/spdk/engine.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (shuo-wu) [reply pkg/spdk/engine.go:791]: I see. Then adding `{UserCreated: false}` would only help us implicitly figure out if a lvol is a replica snapshot/head. It's not important anyway.

**Comment by DamiaSan** [comment]
Codespell is returning the following error:
`Error: ./proto/vendor/protobuf/src/google/protobuf/empty.proto:43: re-use ==> reuse`
What should I do? Modify this file?

**Comment by innobead** [comment]
Codespell errro is fine. Just skip it.

**Comment by DamiaSan** [comment]
> Codespell errro is fine. Just skip it.

So, once you will have resolved the requested changes (I have already addressed all of them), the PR can be merged also with codespell errors?

**Comment by innobead** [comment]
Yes. However, we will fix the codespell setting later.

File: pkg/spdk/replica.go [hunk 7]
```diff
@@ -1427,7 +1440,7 @@ func (r *Replica) RebuildingDstSnapshotCreate(spdkClient *spdkclient.Client, sna
 	}()
 
 	snapLvolName := GetReplicaSnapshotLvolName(r.Name, snapshotName)
-	snapUUID, err := spdkClient.BdevLvolSnapshot(r.rebuildingLvol.UUID, snapLvolName)
+	snapUUID, err := spdkClient.BdevLvolSnapshot(r.rebuildingLvol.UUID, snapLvolName, nil)
```
[the remark this answers is not in the document]
Comment (shuo-wu) [reply pkg/spdk/replica.go:1443]: Remember to update this.
This API will be invoked by the engine [rebuilding here](https://github.com/longhorn/longhorn-spdk-engine/blob/b4718be7c8a94629e374d9a0eda659c515620944/pkg/spdk/engine.go#L982). The engine should check the flag of each snapshot then input the `opts` correct for `RebuildingDstSnapshotCreate` invocation

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: pkg/spdk/replica.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/spdk/replica.go:1443]: Sorry but I don't understand well what you are asking to me. If this API is invoked by the engine during the rebuild, the snapshot is not a user created snapshot ans so there is no necessity to pass opts here.

**Comment by innobead** [comment]
@DamiaSan is this PR ready?

File: pkg/spdk/replica.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (shuo-wu) [reply pkg/spdk/replica.go:1443]: This function copies all snapshots from a healthy replica to the new replica. If a snapshot is requested by users, this snapshot should contain the attribute `userCreated: true` for **all** replicas. 
In other words, the tag `user created` means if the snapshot CR of **the volume** is requested by the users themselves. It does not stand for the snapshot file in a specific replica being created by API `SnapshotCreate` or rebuild from other replicas.

[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/spdk/replica.go:1443]: Ok thanks, I will add this

**Comment by DamiaSan** [comment]
I am going to close this PR and will make a new one.

**Comment by DamiaSan** [comment]
https://github.com/longhorn/longhorn-spdk-engine/pull/115

**Pull request closed without merging**